frame/system/src/offchain.rs

```diff
@@ -17,349 +17,448 @@
-	/// Runtime implementation is free to construct the payload to sign and the signature
-	/// in any way it wants.
-	/// Returns `None` if signed extrinsic could not be created (either because signing failed
-	/// or because of any other runtime-specific reason).
-	fn create_transaction<F: Signer<Self::Public, Self::Signature>>(
-		call: Extrinsic::Call,
-		public: Self::Public,
-		account: T::AccountId,
-		nonce: T::Index,
-	) -> Option<(Extrinsic::Call, Extrinsic::SignaturePayload)>;
+pub enum ForAll {}
+pub enum ForAny {}
+
+pub struct Signer<T: SigningTypes, C: AppCrypto<T::Public, T::Signature>, X = ForAny> {
+	accounts: Option<Vec<T::Public>>,
+	_phantom: sp_std::marker::PhantomData<(X, C)>,
 }
 
-/// A trait responsible for signing a payload using given account.
-///
-/// This trait is usually going to represent a local public key
-/// that has ability to sign arbitrary `Payloads`.
-///
-/// NOTE: Most likely you don't need to implement this trait manually.
-/// It has a blanket implementation for all `RuntimeAppPublic` types,
-/// so it's enough to pass an application-specific crypto type.
-///
-/// To easily create `SignedTransaction`s have a look at the
-/// [`TransactionSubmitter`] type.
-pub trait Signer<Public, Signature> {
-	/// Sign any encodable payload with given account and produce a signature.
-	///
-	/// Returns `Some` if signing succeeded and `None` in case the `account` couldn't
-	/// be used (for instance we couldn't convert it to required application specific crypto).
-	fn sign<Payload: Encode>(public: Public, payload: &Payload) -> Option<Signature>;
+impl<T: SigningTypes, C: AppCrypto<T::Public, T::Signature>, X> Default for Signer<T, C, X> {
+	fn default() -> Self {
+		Self {
+			accounts: Default::default(),
+			_phantom: Default::default(),
+		}
+	}
 }
 
-/// A `Signer` implementation for any `AppPublic` type.
-///
-/// This implementation additionally supports conversion to/from multi-signature/multi-signer
-/// wrappers.
-/// If the wrapped crypto doesn't match `AppPublic`s crypto `None` is returned.
-impl<Public, Signature, TAnyAppPublic> Signer<Public, Signature> for TAnyAppPublic where
-	TAnyAppPublic: RuntimeAppPublic
-		+ AppPublic
-		+ From<<TAnyAppPublic as AppPublic>::Generic>,
-	<TAnyAppPublic as RuntimeAppPublic>::Signature: AppSignature,
-	Signature: From<
-		<<TAnyAppPublic as RuntimeAppPublic>::Signature as AppSignature>::Generic
-	>,
-	Public: TryInto<<TAnyAppPublic as AppPublic>::Generic>
-{
-	fn sign<Payload: Encode>(public: Public, raw_payload: &Payload) -> Option<Signature> {
-		raw_payload.using_encoded(|payload| {
-			let public = public.try_into().ok()?;
-			TAnyAppPublic::from(public).sign(&payload)
-				.map(
-					<<TAnyAppPublic as RuntimeAppPublic>::Signature as AppSignature>
-					 ::Generic::from
-				)
-				.map(Signature::from)
-		})
+impl<T: SigningTypes, C: AppCrypto<T::Public, T::Signature>, X> Signer<T, C, X> {
+	pub fn all_accounts() -> Signer<T, C, ForAll> {
+		Default::default()
```

Comment:
hmm are these mocks?

Reply:
`all_accounts` and `any_account` are part of the Signer interface where you can specify whether you want to sign with all keys or any key. `X` here denotes the mode on which the Signer operates. In `ForAll` case, the signer will operate on all keys (see calls to `for_all`) and the same goes for `for_any`

Reply:
right. Doesn't phantomData have a default? in that case you can skip the `impl Default for Signer`.